PyDynamic/uncertainty/propagate_DFT.py

```diff
@@ -1,776 +1,908 @@
+
+    Calculate the product for V=(V1,V2) and W=(W1,W2). M can be sparse,
+    one-dimensional or a full (quadratic) matrix.
+
+    This is an internal helper function.
+    """
+    if len(M.shape) == 2:
+        assert (M.shape[0] == M.shape[1])
+    assert (M.shape[0] == V.shape[0])
+    assert (V.shape == W.shape)
+    N = V.shape[0] // 2
+    v1 = V[:N]
+    v2 = V[N:]
+    w1 = W[:N]
+    w2 = W[N:]
+    if isinstance(M, sparse.dia_matrix):
+        nrows = M.shape[0]
+        offset = M.offsets
+        diags = M.data
+        A = diags[0][:N]
+        B = diags[1][offset[1]:nrows + offset[1]]
+        D = diags[0][N:]
+        return np.diag(v1 * A * w1 + v2 * B * w1 + v1 * B * w2 + v2 * D * w2)
+    elif len(M.shape) == 1:
+        A = M[:N]
+        D = M[N:]
+        if return_as_matrix:
+            return np.diag(v1 * A * w1 + v2 * D * w2)
+        else:
+            return np.r_[v1 * A * w1 + v2 * D * w2]
+    else:
+        A = M[:N, :N]
+        B = M[:N, N:]
+        D = M[N:, N:]
+        return _prod(v1, _prod(A, w1)) + _prod(v2, _prod(B.T, w1)) + \
+            _prod(v1, _prod(B, w2)) + _prod(v2, _prod(D, w2))
+
+
+def GUM_DFT(x, Ux, N=None, window=None, CxCos=None, CxSin=None, returnC=False,
+            mask=None):
+    """
+    Calculation of the DFT with propagation of uncertainty
+
+    Calculation of the DFT of the time domain signal x and propagation of
+    the squared uncertainty Ux associated with the time domain sequence x to
+    the real and imaginary parts of the DFT of x.
+
+    Parameters
+    ----------
+        x : numpy.ndarray of shape (M,)
+            vector of time domain signal values
+        Ux : numpy.ndarray
+            covariance matrix associated with x, shape (M,M) or noise variance
+            as float
+        N : int, optional
+            length of time domain signal for DFT; N>=len(x)
+        window : numpy.ndarray, optional of shape (M,)
+            vector of the time domain window values
+        CxCos : numpy.ndarray, optional
+            cosine part of sensitivity matrix
+        CxSin : numpy.ndarray, optional
+            sine part of sensitivity matrix
+        returnC : bool, optional
+            if true, return sensitivity matrix blocks for later use
+        mask: ndarray of dtype bool
+            calculate DFT values and uncertainties only at those frequencies
+            where mask is `True`
+
+    Returns
+    -------
+        F : numpy.ndarray
+            vector of complex valued DFT values or of its real and imaginary
+            parts
+        UF : numpy.ndarray
+            covariance matrix associated with real and imaginary part of F
+
+    References
+    ----------
+        * Eichstädt and Wilkens [Eichst2016]_
+    """
+    L = 0
+    if isinstance(window, np.ndarray):
+        x, Ux = _apply_window(x, Ux,
+                              window)  # apply the chose window for the
+        # application of the FFT
```

Comment:
These linebreaks do not contribute to the overall readability in my opinion. I would move the comment above the line of code.

Reply:
Agree!